### PR TITLE
Refactored desig to support F7 with different desig reg addresses

### PR DESCRIPTION
**Pull request by matthewlai**
Refactored desig to support F7 having different reg addresses depending on device ID in DBGMCU

### CODE REVIEW

File: lib/stm32/common/desig_common_all.c [hunk 1]
```diff
@@ -19,35 +19,22 @@
...
-	len = (2 * sizeof(dev_id_buf) < string_len) ?
-		2 * sizeof(dev_id_buf) : string_len - 1;
+	len = (2 * sizeof(uid_buf) < string_len) ?
+		2 * sizeof(uid_buf) : string_len - 1;
 
 	for (i = 0; i < len; i += 2) {
-		string[i]     = chars[(device_id[i / 2] >> 4) & 0x0F];
-		string[i + 1] = chars[(device_id[i / 2] >> 0) & 0x0F];
+		string[i]     = chars[(uid[i / 2] >> 4) & 0x0F];
+		string[i + 1] = chars[(uid[i / 2] >> 0) & 0x0F];
```
Comment (karlp) [inline lib/stm32/common/desig_common_all.c]: this is all gratuitous in this function, you can just drop it.

File: lib/stm32/common/desig_common_all.c [hunk 2]
```diff
@@ -66,15 +53,18 @@ void desig_get_unique_id_as_string(char *string,
...
-	serial[2] = id[9];
-	serial[3] = id[8] + id[0];
-	serial[4] = id[7];
-	serial[5] = id[6];
+	serial[0] = uid[11];
+	serial[1] = uid[10] + uid[2];
+	serial[2] = uid[9];
+	serial[3] = uid[8] + uid[0];
+	serial[4] = uid[7];
+	serial[5] = uid[6];
```
Comment (karlp) [inline lib/stm32/common/desig_common_all.c]: same here, it might be "nice" but it just makes churn for no real particular reason in unrelated code.

File: lib/stm32/f7/desig.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+      flash_size_base = (uint32_t*)DESIG_FLASH_SIZE_BASE_451;
+      break;
+    case 0x0452:
+      flash_size_base = (uint32_t*)DESIG_FLASH_SIZE_BASE_452;
+      break;
+  }
+
+  if (!flash_size_base) {
+    /* We don't know the address for this device. Hang here to help debugging. */
+    for(;;);
```
Comment (karlp) [inline lib/stm32/f7/desig.c]: perhaps nicer to return 0?  we have a cm3_assert you could use too, cm3_assert_not_reached iirc, or similar.

**Comment by karlp** [comment]
I like what you've done here.  we should continue with it, to fix existing bugs such as https://github.com/libopencm3/libopencm3/issues/234 (and iirc, it's also a problem within some other families too, not just l1 and f7)

File: lib/stm32/common/desig_common_all.c (hunk 1 above)
Comment (matthewlai) [reply lib/stm32/common/desig_common_all.c]: I thought it would be nice to change the names here, since "device ID" actually means something else (the device ID field in DBGMCU). I feel it would be especially confusing to anyone tracing this code now since that other device ID is actually used to decide the address now. I can revert them if you want, though. Is there a benefit to not changing this? It's an internal name that no one should really care about?

File: lib/stm32/common/desig_common_all.c (hunk 2 above)
Comment (matthewlai) [reply lib/stm32/common/desig_common_all.c]: This is just to make them match the above.

File: lib/stm32/f7/desig.c (hunk 3 above)
Comment (matthewlai) [reply lib/stm32/f7/desig.c]: I thought about returning 0, but thought it would be nicer to fail loudly rather than silently. What do you think?

cm3_assert_not_reached() is nice. I'll change it tonight when I'm back home.

File: lib/stm32/common/desig_common_all.c (hunk 1 above)
Comment (karlp) [reply lib/stm32/common/desig_common_all.c]: yeah, I mean, whatever, it's just generally bad form to make unrelated changes.  it makes the diffs much bigger and harder to review, and we end up wasting time on this sort of cruft instead of the issues itself.

Comment (matthewlai) [reply lib/stm32/common/desig_common_all.c]: Ah I had these changes in a separate commit before, but squashed them for the PR. Will keep them separate next time.

File: lib/stm32/f7/desig.c (hunk 3 above)
Comment (matthewlai) [reply lib/stm32/f7/desig.c]: Done.

**Comment by karlp** [comment]
merged thanks! (I fixed your whitespace for you)
